# Choosing Machine Learning Model

# Regression Model

In [3]:
# standerd imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
%matplotlib inline

![Select Model](https://scikit-learn.org/1.3/_static/ml_map.png)

In [ ]:
# how to select a model
# Why, How, when
from sklearn.datasets import fetch_california_housing
housing = fetch_california_housing()
housing

In [6]:
df_1 = pd.read_csv('housing.csv')
df_1

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     PTRATIO       B  LSTAT  target  
0       15.3  396.90   4.98    24.0  
1       17.8  396.90   9.14    21.6  
2       17.8  392.83   4.03    34.7  
3       18.7  394.63   2.94    33.4  
4       18.7  396.90   5.33    36.2  
..       ...     ...    ...     ...  
501     21.0  391.99   9.67    22.4  
502     21.0  396.90   9.08    20.6  
503     21.0  396.90   5.64    23.9  
504     21.0  393.45   6.48    22.0  
505     21.0  396.90   7.88    11.9  

[506 rows x 14 columns]

### 1. when we have to predict a number its a= reggression problem
### 2. Labeled data? = yes
### 3. Number of samples = 506

In [9]:
# lets try ridge reggression model
from sklearn.linear_model import Ridge
# to reproduce same results 
np.random.seed(1)
# create data
X = df_1.drop('target', axis = 1)
y = df_1['target']
# Model split test and train
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=.20)
# instantiate the Ridge Model
reg_model = Ridge()
reg_model.fit(X_train, y_train)
reg_model.score(X_test, y_test)


0.7655800611077146

In [12]:
# ensemble Random Fores
from sklearn.ensemble import RandomForestRegressor
# to reproduce same results 
np.random.seed(1)
# create data
X = df_1.drop('target', axis = 1)
y = df_1['target']
# Model split test and train
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=.20)
# instantiate the Ridge Model
rmf_model = RandomForestRegressor()
rmf_model.fit(X_train, y_train)
rmf_model.score(X_test, y_test)


0.9124687687774722

# Classification Model

In [15]:
# choosing Estimation for Classification
heart_disease = pd.read_csv('heart.csv')
heart_disease


age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0     63    1   3     145   233    1        0       150     0      2.3    0   
1     37    1   2     130   250    0        1       187     0      3.5    0   
2     41    0   1     130   204    0        0       172     0      1.4    2   
3     56    1   1     120   236    0        1       178     0      0.8    2   
4     57    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output  
0      0      1       1  
1      0      2       1  
2      0      2       1  
3      0      2       1  
4      0      2       1  
..   ...    ...     ...  
298    0      3       0  
299    0      3       0  
300    2      3       0  
301    1      3       0  
302    1      2       0  

[303 rows x 14 columns]

In [40]:
# SVM 
from sklearn.svm import LinearSVC
# to reproduce same results 
np.random.seed(1)
# create data
X = heart_disease.drop('output', axis = 1)
y = heart_disease['output']
# Model split test and train
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=.20)
# instantiate the SVC Model
svc_model = LinearSVC()
svc_model.fit(X_train, y_train)
svc_model.score(X_test, y_test)


0.7377049180327869

In [41]:
# linera reggressor classifier
from sklearn.ensemble import RandomForestClassifier
# to reproduce same results 
np.random.seed(1)
# create data
X = heart_disease.drop('output', axis = 1)
y = heart_disease['output']
# Model split test and train
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=.20)
# instantiate the SVC Model
rfc_model = RandomForestClassifier()
rfc_model.fit(X_train, y_train)
rfc_model.score(X_test, y_test)


0.7868852459016393